beacon_node/store/src/hot_cold_store.rs

```diff
@@ -2861,10 +2861,20 @@ pub fn migrate_database<E: EthSpec, Hot: ItemStore<E>, Cold: ItemStore<E>>(
             hot_db_ops.push(StoreOp::DeleteExecutionPayload(block_root));
         }
 
-        if slot % E::slots_per_epoch() != 0 {
-            non_epoch_boundary_blocks.push((slot, block_root))
+        // At a missed slot, `state_root_iter` will return the block root
+        // from the previous non-missed slot. This ensures that the block root at an
+        // epoch boundary is always a checkpoint block root. We keep track of block roots
+        // at epoch boundaries by storing them in the `epoch_boundary_blocks` hash set.
+        // We then ensure that block roots at the epoch boundary aren't included in the
+        // `non_checkpoint_block_roots` hash set.
+        if slot % E::slots_per_epoch() == 0 {
+            epoch_boundary_blocks.insert(block_root);
         } else {
-            epoch_boundary_blocks.push((slot, block_root))
+            non_checkpoint_block_roots.insert(block_root);
+        }
+
+        if epoch_boundary_blocks.contains(&block_root) {
+            non_checkpoint_block_roots.remove(&block_root);
```

Comment:
Much cleaner <3 

The intended logic is:

_no skipped slot_
- Slot 2/1  root C  : Add C to non_checkpoint_block_roots
- Slot 2/0  root B  : Add B to epoch_boundary_blocks
- Slot 1/31 root A  : Add A to non_checkpoint_block_roots

_slot 31 skipped_
- Slot 2/1  root C  : Add C to non_checkpoint_block_roots
- Slot 2/0  root A  : Add A to epoch_boundary_blocks
- Slot 1/31 root A  : Add A to non_checkpoint_block_roots, remove A from non_checkpoint_block_roots

_slot 0 skipped_
- Slot 2/1  root B  : Add B to non_checkpoint_block_roots
- Slot 2/0  root B  : Add B to epoch_boundary_blocks, remove B from non_checkpoint_block_roots
- Slot 1/31 root A  : Add A to non_checkpoint_block_roots